src/dirent.rs

```diff
@@ -0,0 +1,40 @@
+use {Result, Error, Errno};
+use libc::{self, DIR};
+use std::os::unix::io::RawFd;
+use errno;
+
+pub struct Dir {
+    pub handle: *mut DIR,
+}
+
+impl Dir {
+    pub fn read(&mut self) -> Result<Option<&libc::dirent>> {
+        let dirent = unsafe {
+            Errno::clear();
+            libc::readdir(self.handle)
+        };
+        if dirent.is_null() {
+            match Errno::last() {
+                errno::UnknownErrno => Ok(None),
+                _ => Err(Error::last().into()),
+            }
+        } else {
+            Ok(Some(unsafe{&*dirent}))
```

Comment:
Could you annotate the lifetime in the signature so we're sure that this reference has the correct lifetime:
```
pub fn read<'a>(&'a mut self) -> Result<Option<&'a libc::dirent>>